{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following problem sets have been taken from the following webpage:\n",
    "\n",
    "http://www2.mpia-hd.mpg.de/~robitaille/PY4SCI_SS_2014/\n",
    "\n",
    "For a newer version, check out:\n",
    "\n",
    "http://www2.mpia-hd.mpg.de/~robitaille/PY4SCI_SS_2015/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Part 1\n",
    "\n",
    "# Question 1\n",
    "\n",
    "def reverse_complement(DNA):\n",
    "    str = \"\"\n",
    "    for i in range(len(DNA)-1,-1,-1):\n",
    "        if DNA[i] == 'A':\n",
    "            str = str + 'T'\n",
    "        elif DNA[i] == 'T':\n",
    "            str = str + 'A'\n",
    "        elif DNA[i] == 'C':\n",
    "            str = str + 'G'\n",
    "        elif DNA[i] == 'G':\n",
    "            str = str + 'C'\n",
    "    return str\n",
    "\n",
    "a = \"ACTTGCTATCGTAGCATGGGCGTATTATTAGGCTAGTCGGATTGCGATCGGATCGATTGCA\"\n",
    "\n",
    "print(reverse_complement(a))\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Question 2\n",
    "\n",
    "def dna_to_mrna(DNA):\n",
    "    str = \"\"\n",
    "    for i in range(len(DNA)):\n",
    "        if DNA[i] == 'T':\n",
    "            str = str + 'U'\n",
    "        else:\n",
    "            str = str + DNA[i]\n",
    "    return str\n",
    "\n",
    "a = \"ATCGCGAT\"\n",
    "\n",
    "print(dna_to_mrna(a))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Part 2\n",
    "\n",
    "# Question 3\n",
    "\n",
    "# dictionary relating codons to corresponding amino acids\n",
    "codons = {}\n",
    "f = open('data/problem_1_codons.txt', 'r')\n",
    "for line in f:\n",
    "    line = line.strip()\n",
    "    columns = line.split()\n",
    "    codons[columns[0]] = columns [1]\n",
    "    \n",
    "# converting dna sequence to mrna\n",
    "def dna_to_mrna(DNA):\n",
    "    str = \"\"\n",
    "    for i in range(len(DNA)):\n",
    "        if DNA[i] == 'T':\n",
    "            str = str + 'U'\n",
    "        else:\n",
    "            str = str + DNA[i]\n",
    "    return str\n",
    "\n",
    "# converting mrna sequence - each codon - 3 nucleotides into protein (amino acids)\n",
    "def mrna_to_protein(mrna):\n",
    "    amino = \"\"\n",
    "    for i in range(0,len(mrna),3):\n",
    "        seq = mrna[i:i+3]\n",
    "        if codons[seq] == \"Stop\":\n",
    "            break\n",
    "        else:\n",
    "            amino = amino + codons[seq]\n",
    "    return amino\n",
    "\n",
    "# converting a dna sequence into mrna and then into protein\n",
    "def dna_to_protein(dna):\n",
    "    mrna = dna_to_mrna(dna)\n",
    "    protein = mrna_to_protein(mrna)\n",
    "    return protein\n",
    "        \n",
    "a = \"AATCTCTACGGAAGTAGGTCAGTACTGATCGATCAGTCGATCGGGCGGCGATTTCGATCTGATTGTACGGCGGGCTAG\"        \n",
    "        \n",
    "print \"Corresponding protein is:\",dna_to_protein(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sequence_08 : DPLRFSVGTERRGSALVFFILGGI\n",
      "sequence_15 : GQQHSMGTLYKM\n",
      "sequence_18 : LELFGEDDTVE\n",
      "sequence_19 : GLLPDSYRVSLSLQRTTLCLTIVCSGLCGLPDLMNLESRFHMQSVTF\n",
      "sequence_01 : YKFAGYQAIIGPRPATSKGLRLTAIQWRHLRWPIICIPEEHWY\n",
      "sequence_12 : RPFPAVCRIKFFWPYISIKEGSANASNTTNC\n",
      "sequence_03 : RPERFGASRYTICVHVSDNPPVSTLLSKEQTGGA\n",
      "sequence_02 : RSFCTGHPSYCRLFFGRKMQSSFAPR\n",
      "sequence_05 : TLITPLGYPGRIGELRGLVYVRGHHQVPSMCTYS\n",
      "sequence_04 : CMSSRNAITGCITCVGHIMGVRLYD\n",
      "sequence_07 : CVARTGWQMRTAPVSFAPSPPINAPKRTPSIRWMSFLIP\n",
      "sequence_06 : CIEPELQATRVKATFSFASCPSDMFSKSPL\n",
      "sequence_16 : AIRCRVAIESDLTVVARLVYIVPCTPSDSRNYPNSSGFKPISKTQNL\n",
      "sequence_09 : LTFEFLARGDQFSTWWL\n",
      "sequence_10 : PSENPNVSSRPPPMIFTDFT\n",
      "sequence_17 : KCRRFRQTTTYKPCSCSVIRCTREQQD\n",
      "sequence_13 : LKGIIRCSDSNCSSPIFGPSQEIQM\n",
      "sequence_11 : RALGATVNMLGYHVFPN\n",
      "sequence_14 : IMTNPTLNKDTA\n"
     ]
    }
   ],
   "source": [
    "# Question 4\n",
    "\n",
    "# dictionary relating codons to corresponding amino acids\n",
    "codons = {}\n",
    "f1 = open('data/problem_1_codons.txt', 'r')\n",
    "for line in f1:\n",
    "    line = line.strip()\n",
    "    columns = line.split()\n",
    "    codons[columns[0]] = columns [1]\n",
    "    \n",
    "# converting dna sequence to mrna\n",
    "def dna_to_mrna(DNA):\n",
    "    str = \"\"\n",
    "    for i in range(len(DNA)):\n",
    "        if DNA[i] == 'T':\n",
    "            str = str + 'U'\n",
    "        else:\n",
    "            str = str + DNA[i]\n",
    "    return str\n",
    "\n",
    "# converting mrna sequence - each codon - 3 nucleotides into protein (amino acids)\n",
    "def mrna_to_protein(mrna):\n",
    "    amino = \"\"\n",
    "    for i in range(0,len(mrna),3):\n",
    "        seq = mrna[i:i+3]\n",
    "        if codons[seq] == \"Stop\":\n",
    "            break\n",
    "        else:\n",
    "            amino = amino + codons[seq]\n",
    "    return amino\n",
    "\n",
    "# converting a dna sequence into mrna and then into protein\n",
    "def dna_to_protein(dna):\n",
    "    mrna = dna_to_mrna(dna)\n",
    "    protein = mrna_to_protein(mrna)\n",
    "    return protein\n",
    "\n",
    "# reading from FASTA file and making dictionary\n",
    "def read_fasta(filename):\n",
    "    e = {}\n",
    "    f2 = open(filename, 'r')\n",
    "    label = \"\"\n",
    "    seq = \"\"\n",
    "    for line in f2:\n",
    "        l = line.strip()\n",
    "        if l[0] == '>':\n",
    "            e[label] = seq\n",
    "            label = l[1:]\n",
    "            seq = \"\"\n",
    "        else:\n",
    "            seq = seq + l\n",
    "    \n",
    "    try:\n",
    "        del e[\"\"]\n",
    "    except KeyError:\n",
    "        pass\n",
    "    \n",
    "    return e\n",
    "\n",
    "filename = 'data/problem_1_question_4.fasta'\n",
    "dna_dict = read_fasta(filename)\n",
    "\n",
    "for key in dna_dict:\n",
    "    print key, \":\" , dna_to_protein(dna_dict[key])\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GTACTAGCCGATCAGCTAGCTACGATCGTACGTTAA\n"
     ]
    }
   ],
   "source": [
    "# Part 3\n",
    "\n",
    "# Question 5\n",
    "\n",
    "# reading from FASTA file and making dictionary\n",
    "def read_fasta(filename):\n",
    "    e = {}\n",
    "    f2 = open(filename, 'r')\n",
    "    label = \"\"\n",
    "    seq = \"\"\n",
    "    for line in f2:\n",
    "        l = line.strip()\n",
    "        if l[0] == '>':\n",
    "            e[label] = seq\n",
    "            label = l[1:]\n",
    "            seq = \"\"\n",
    "        else:\n",
    "            seq = seq + l\n",
    "    \n",
    "    try:\n",
    "        del e[\"\"]\n",
    "    except KeyError:\n",
    "        pass\n",
    "    \n",
    "    return e\n",
    "\n",
    "# finding the consensus sequence\n",
    "def consensus_sequence(dna_dict):\n",
    "    \n",
    "    length = 0\n",
    "    \n",
    "    for key in dna_dict:\n",
    "        seq = dna_dict[key]\n",
    "        length = len(seq)\n",
    "        break\n",
    "    \n",
    "    consensus = \"\"\n",
    "    index_label = {0 : 'A' , 1 : 'T' , 2 : 'C' , 3 : 'G'}\n",
    "    \n",
    "    for i in range(length):\n",
    "        count = [0,0,0,0]\n",
    "        # indices 0,1,2,3 correspond to A,T,C and G respectively\n",
    "        for key in dna_dict:\n",
    "            seq = dna_dict[key]\n",
    "            if seq[i] == 'A':\n",
    "                count[0] = count[0] + 1\n",
    "            elif seq[i] == 'T':\n",
    "                count[1] = count[1] + 1\n",
    "            elif seq[i] == 'C':\n",
    "                count[2] = count[2] + 1\n",
    "            elif seq[i] == 'G':\n",
    "                count[3] = count[3] + 1\n",
    "                \n",
    "        max_index = count.index(max(count))\n",
    "        consensus = consensus + index_label[max_index]\n",
    "        \n",
    "    return consensus\n",
    "        \n",
    "filename = 'data/problem_1_question_6.fasta'\n",
    "dna_dict = read_fasta(filename)\n",
    "\n",
    "print consensus_sequence(dna_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
